dashboard/test/controllers/omniauth_callbacks_controller_test.rb

```diff
@@ -537,6 +537,95 @@ class OmniauthCallbacksControllerTest < ActionController::TestCase
+    #Given I have a Google-Code.org account
+    user = create :student, :google_sso_provider
+    user_auth = user.authentication_options.find_by_credential_type(AuthenticationOption::GOOGLE)
+
+    #Generate token from 6 minutes ago
+    secret_code = Encryption.encrypt_string_utf8(
+      (Time.now - 6.minutes).strftime('%Y%m%dT%H%M%S%z') + user_auth['authentication_id'] + user_auth['credential_type']
+    )
+
+    # Go to function
+    post :maker_google_oauth2, params: {secret_code: secret_code}
+
+    assert_template 'maker/login_code'
+
+    # Check that flash alert displays desired message
+    expected_error = I18n.t('maker.google_oauth.error_token_expired')
+    assert_select ".container .alert-danger", expected_error
+  end
+
+  test 'maker_google_oauth2: returns an error if provider is incorrect' do
+    #Given I have a Google-Code.org account
+    user = create :student, :google_sso_provider
+    user_auth = user.authentication_options.find_by_credential_type(AuthenticationOption::GOOGLE)
+
+    #Generate token with incorrect provider
+    secret_code = Encryption.encrypt_string_utf8(
+      Time.now.strftime('%Y%m%dT%H%M%S%z') + user_auth['authentication_id'] + "Clever"
+    )
+
+    # Go to function
+    post :maker_google_oauth2, params: {secret_code: secret_code}
+
+    assert_template 'maker/login_code'
+
+    # Check that flash alert displays desired message
+    expected_error = I18n.t('maker.google_oauth.error_wrong_provider')
+    assert_select ".container .alert-danger", expected_error
+  end
+
+  test 'maker_google_oauth2: returns an error if user if is invalid' do
+    #Given I have a Google-Code.org account
+    user = create :student, :google_sso_provider
+    user_auth = user.authentication_options.find_by_credential_type(AuthenticationOption::GOOGLE)
+
+    #Generate token with corrupted user id
+    secret_code = Encryption.encrypt_string_utf8(
+      Time.now.strftime('%Y%m%dT%H%M%S%z') + user_auth['authentication_id'] + "test" + user_auth['credential_type']
+    )
+
+    # Go to function
+    post :maker_google_oauth2, params: {secret_code: secret_code}
+
+    assert_template 'maker/login_code'
+
+    # Check that flash alert displays desired message
+    expected_error = I18n.t('maker.google_oauth.error_invalid_user')
+    assert_select ".container .alert-danger", expected_error
+  end
+
+  test 'maker_google_oauth2: logs in user if valid token' do
+    #Given I have a Google-Code.org account
+    user = create :student, :google_sso_provider
+    user_auth = user.authentication_options.find_by_credential_type(AuthenticationOption::GOOGLE)
+
+    #Generate token
+    secret_code = Encryption.encrypt_string_utf8(
+      Time.now.strftime('%Y%m%dT%H%M%S%z') + user_auth['authentication_id'] + user_auth['credential_type']
+    )
+
+    #Go to function
+    post :maker_google_oauth2, params: {secret_code: secret_code}
+
+    #Then I am signed in
+    user.reload
```

Comment:
nit: we don't need to reload the user here. reload is used when you update a model that you already have a reference to:

```
user = create :student, name: "Old Name"
user.update(name: "New Name!!!")

user.name => "Old Name"
user.reload
user.name => "New Name!!!"
```